File: loopx/control_plane/todos/decision_scope.ts
/** Read-only decision dependency rules over one complete source snapshot.
 * A consistent dependency is not approval, a lease, or a mutation receipt. */
import type {JsonObject} from "../effect_program.ts";
import {requireJsonObject, optionalNonEmptyString, requireBoolean, requireInteger} from "../runtime_decode.ts";
import {gateAddressesAgent} from "./gate_scope.ts";

export const DECISION_SCOPE_REQUEST_SCHEMA = "todo_decision_scope_request_v0";
const RANK: Readonly<Record<string, number>> = {action: 0, lane: 1, goal: 2, project: 3, global: 4};
const KINDS = new Set(["private_read", "write_scope", "resource", "production", "public_claim", "direction", "other"]);
const object = (value: unknown): value is JsonObject => value !== null && typeof value === "object" && !Array.isArray(value);
const rows = (value: unknown): JsonObject[] => {
  if (!Array.isArray(value)) throw new TypeError("decision scope rows must be an array");
  return value.map(item => requireJsonObject(item, "decision scope row"));
};
const text = (value: unknown): string | null => typeof value === "string" && value ? value : null;

export function decisionScopeCovers(gate: unknown, required: unknown): boolean {
  if (!object(gate) || !object(required)) return false;
  const valid = (scope: JsonObject) => KINDS.has(String(scope.kind)) && Object.hasOwn(RANK, String(scope.granularity)) &&
    typeof scope.scope_key === "string" && /^(?:\*|[a-z0-9][a-z0-9_.:@*/-]{0,95})$/u.test(scope.scope_key);
  return valid(gate) && valid(required) && gate.kind === required.kind &&
    (gate.scope_key === "*" || gate.scope_key === required.scope_key) && RANK[String(gate.granularity)] >= RANK[String(required.granularity)];
}

function addressed(gate: JsonObject, agent: string | null): boolean {
  return gateAddressesAgent({global: gate.global_gate === true, blocks: text(gate.blocks_agent), claim: text(gate.claimed_by)}, agent);
}

export function scopeStandingAuthority(value: unknown, agent: string | null): JsonObject | null {
  if (!object(value) || !Array.isArray(value.entries)) return null;
  const entries = rows(value.entries).filter(item => addressed(item, agent));
  const conflicts = rows(value.conflicts ?? []).filter(item => addressed(item, agent));
  if (!entries.length && !conflicts.length) return null;
  return {schema_version: "standing_decision_authority_v0", agent_id: agent,
    active_count: entries.filter(e => e.active === true).length,
    inactive_count: entries.filter(e => e.active !== true).length, entries,
    ...(conflicts.length ? {conflicts, conflict_count: conflicts.length} : {})};
}

export function decisionScopeRelation(gate: JsonObject, item: JsonObject): JsonObject | null {
  const scope = gate.decision_scope;
  if (!object(scope)) return null;
  const required = rows(item.required_decision_scopes ?? []);
  const matched = required.find(value => decisionScopeCovers(scope, value));
  return {schema_version: "decision_scope_relation_v0", source: "decision_scope",
    state: matched ? "gate_covers_action" : "independent",
    ...(!required.length ? {reason: "agent_todo_has_no_required_decision_scopes"} : {}),
    gate_todo_id: gate.todo_id ?? null, agent_todo_id: item.todo_id ?? null,
    decision_scope: scope, ...(matched ? {matched_required_decision_scope: matched} : {}), required_decision_scopes: required};
}

export function exactTodoGateRelation(gate: JsonObject, item: JsonObject): JsonObject | null {
  const target = text(gate.unblocks_todo_id);
  if (!target) return null;
  return {schema_version: "todo_gate_relation_v0", source: "unblocks_todo_id",
    state: target === item.todo_id ? "gate_targets_todo" : "independent",
    gate_todo_id: gate.todo_id ?? null, target_todo_id: target, agent_todo_id: item.todo_id ?? null};
}

export function todoGateRelation(gate: JsonObject, item: JsonObject): JsonObject | null {
  const exact = exactTodoGateRelation(gate, item), scope = decisionScopeRelation(gate, item);
  if (exact?.state === "independent" && scope?.state === "gate_covers_action") {
    return {schema_version: "todo_gate_relation_v0", source: "unblocks_todo_id+decision_scope",
      state: "projection_repair_required", reason: "decision_scope_covers_agent_todo_but_unblocks_todo_targets_different_todo",
      gate_todo_id: exact.gate_todo_id, target_todo_id: exact.target_todo_id,
      agent_todo_id: exact.agent_todo_id, exact_todo_relation: exact, decision_scope_relation: scope};
  }
  if (exact?.state === "gate_targets_todo") return {...exact, ...(scope ? {decision_scope_relation: scope} : {})};
  if (scope?.state === "gate_covers_action") return {...scope, ...(exact ? {exact_todo_relation: exact} : {})};
  return exact ? {...exact, ...(scope ? {decision_scope_relation: scope} : {})} : scope;
}

/** Legacy action labels are exact keys, not natural-language permission rules.
 * Missing scope is insufficient evidence for safe bypass, never a global gate. */
export function fallbackGateRelation(gate: JsonObject, item: JsonObject): JsonObject {
  if (gate.global_gate === true) return {source: "global_gate", state: "gate_covers_action"};
  const explicit = todoGateRelation(gate, item);
  if (explicit) return explicit;
  const action = (row: JsonObject) => typeof row.action_kind === "string" ? row.action_kind.trim().toLowerCase() : "";
  const a = action(gate), b = action(item);
  return {schema_version: "todo_gate_relation_v0", gate_todo_id: gate.todo_id ?? null,
    agent_todo_id: item.todo_id ?? null,
    source: a && a === b ? "legacy_exact_action_kind" : "missing_dependency_scope",
    state: a && a === b ? "gate_covers_action" : "projection_repair_required",
    ...(!a || a !== b ? {reason: "safe_fallback_requires_explicit_dependency_scope"} : {})};
}

/** Select only from already evaluated lanes. Never re-evaluate waits against a
 * compact list or refill an authoritative empty capability result from backlog.
 * Return source positions, leaving display compaction and wording to adapters. */
export function selectScopedGateFallback(request: JsonObject): JsonObject | null {
  const agent = optionalNonEmptyString(request.agent_id, "agent_id");
  const debt = requireBoolean(request.monitor_debt_backoff_active, "monitor_debt_backoff_active");
  const allowUnrelated = requireBoolean(request.allow_unrelated_gate, "allow_unrelated_gate");
  const gates = rows(request.gates).map((gate, index) => ({gate, index})).filter(({gate}) =>
    gate.done !== true && ["open", "blocked"].includes(String(gate.status || "open")) &&
    gate.archive_state !== "archive" && (!agent || addressed(gate, agent)));
  if (!gates.length) return null;
  const source = rows(request.candidates).map((item, index) => ({item, index,
    priority: requireInteger(item.priority_rank, "priority_rank"),
    persisted: requireInteger(item.persisted_index, "persisted_index")}));
  const seen = new Set<string>();
  const candidates = source.filter(({item}) => {
    const id = text(item.todo_id);
    if (id && seen.has(id)) return false;
    if (id) seen.add(id);
    const deferred = item.status === "deferred" && item.resume_ready === true;
    return item.archive_state !== "archive" && ((item.status === "open" && item.done !== true) || deferred) &&
      item.removed !== true && (!agent || ((!item.claimed_by || item.claimed_by === agent) &&
        (!item.bound_agent || item.bound_agent === agent) &&
        !(Array.isArray(item.excluded_agents) && item.excluded_agents.includes(agent))));
  });
  candidates.sort((a, b) => a.priority - b.priority ||
    (debt ? Number(a.item.task_class !== "advancement_task") - Number(b.item.task_class !== "advancement_task") : 0) ||
    a.persisted - b.persisted || a.index - b.index);
  const blocked: JsonObject[] = [];
  let selected: typeof candidates[number] | undefined;
  let blockingGate: typeof gates[number] | undefined;
  for (const candidate of candidates) {
    const match = gates.map(g => ({...g, relation: fallbackGateRelation(g.gate, candidate.item)}))
      .find(g => g.relation.state !== "independent");
    if (match) {
      blockingGate ??= match;
      blocked.push({candidate_index: candidate.index, gate_index: match.index, relation: match.relation});
    } else selected ??= candidate;
  }
  if (!selected || (!blockingGate && !allowUnrelated)) return null;
  const surface = blockingGate ?? gates[0]!;
  return {schema_version: "scoped_gate_fallback_selection_v0",
    selected_index: selected.index, gate_index: surface.index, blocked: blocked.slice(0, 3), blocked_count: blocked.length,
    has_blocking_gate: blockingGate !== undefined,
    selected_relation: fallbackGateRelation(surface.gate, selected.item),
    deferred_replan: selected.item.status === "deferred" && selected.item.resume_ready === true};
}

function open(items: JsonObject[]): JsonObject[] {
  return items.filter(item => item.done !== true && ["open", "blocked"].includes(String(item.status || "open")));
}
function scopeId(scope: JsonObject): string { return `${scope.kind}:${scope.granularity}:${scope.scope_key}`; }

export function decisionScopeConsistency(request: JsonObject): JsonObject {
  const agent = optionalNonEmptyString(request.agent_id, "agent_id");
  const agents = rows(request.agent_items), users = open(rows(request.user_items));
  const items = open(agents), gates = users.filter(item => item.is_gate === true), actions = users.filter(item => item.is_gate !== true);
  if (!Array.isArray(request.registered_agents) || request.registered_agents.some(id => typeof id !== "string")) {
    throw new TypeError("registered agents must be normalized strings");
  }
  const registered = [...new Set(request.registered_agents as string[])].sort();
  const errors: JsonObject[] = [];
  if (registered.length > 1) for (const gate of gates) {
    if (gate.global_gate !== true && !gate.blocks_agent) errors.push({reason_code: "multi_agent_user_gate_missing_scope",
      user_todo_id: gate.todo_id ?? null, registered_agent_ids: registered});
  }
  let checked = 0, terminals = 0, approvals = 0;
  for (const item of items) {
    const claim = text(item.claimed_by);
    if (claim && agent && claim !== agent) continue;
    const owner = agent || claim;
    const authority = scopeStandingAuthority(request.standing_authority, owner);
    for (const scope of rows(item.required_decision_scopes ?? [])) {
      checked++;
      const matching = gates.filter(gate => decisionScopeCovers(gate.decision_scope, scope));
      const compatible = matching.filter(gate => addressed(gate, owner));
      // Exact targeting is a restriction, not a hint that a broad scope can erase.
      const conflicting = compatible.filter(gate => todoGateRelation(gate, item)?.state === "projection_repair_required");
      if (conflicting.length) {
        errors.push({reason_code: "required_decision_scope_target_mismatch", agent_todo_id: item.todo_id ?? null,
          required_scope: scopeId(scope), related_user_todo_ids: conflicting.map(g => g.todo_id).filter(Boolean)});
        continue;
      }
      if (compatible.length) continue;
      if (rows(authority?.entries ?? []).some(entry => entry.active === true && decisionScopeCovers(entry.decision_scope, scope))) {
        approvals++; continue;
      }
      const outcomes = rows(item.decision_scope_outcomes ?? []).filter(outcome => ["reject", "cancel"].includes(String(outcome.outcome)) &&
        decisionScopeCovers(outcome.decision_scope, scope));
      if (outcomes.length) {
        terminals++;
        if ((item.status || "open") !== "blocked") errors.push({reason_code: "terminal_decision_outcome_target_not_blocked",
          agent_todo_id: item.todo_id ?? null, required_scope: scopeId(scope), related_user_todo_ids: outcomes.map(o => o.source_todo_id)});
        continue;
      }
      const conflicts = rows(authority?.conflicts ?? []).filter(entry => decisionScopeCovers(entry.decision_scope, scope));
      const matchingActions = actions.filter(action => decisionScopeCovers(action.decision_scope, scope));
      let reason = "dangling_required_decision_scope", related: unknown[] = [];
      if (conflicts.length) {
        reason = "standing_decision_order_unresolved";
        related = [...new Set(conflicts.flatMap(c => Array.isArray(c.source_todo_ids) ? c.source_todo_ids.filter(id => typeof id === "string") : []))].sort();
      } else if (matchingActions.length) {
        reason = "non_blocking_user_action_scope_collision"; related = matchingActions.map(a => a.todo_id).filter(Boolean);
      } else if (matching.length) {
        reason = "required_decision_scope_gate_owner_mismatch"; related = matching.map(g => g.todo_id).filter(Boolean);
      }
      errors.push({reason_code: reason, agent_todo_id: item.todo_id ?? null, required_scope: scopeId(scope),
        related_user_todo_ids: related as string[]});
    }
  }
  return {schema_version: "required_decision_scope_consistency_v0", ok: !errors.length,
    status: errors.length ? "projection_repair_required" : "consistent", agent_id: agent,
    checked_agent_todo_count: items.length, checked_required_scope_count: checked,
    terminal_outcome_count: terminals, standing_authority_match_count: approvals, errors};
}

export function evaluateDecisionScope(value: unknown): JsonObject {
  const request = requireJsonObject(value, "decision scope request");
  if (request.schema_version !== DECISION_SCOPE_REQUEST_SCHEMA) throw new TypeError("decision scope request schema mismatch");
  let result: JsonObject | boolean | null | (JsonObject | null)[][];
  switch (request.operation) {
    case "fallback": result = selectScopedGateFallback(request); break;
    case "consistency": result = decisionScopeConsistency(request); break;
    case "standing": result = scopeStandingAuthority(request.authority, optionalNonEmptyString(request.agent_id, "agent_id")); break;
    case "covers": result = decisionScopeCovers(request.gate_scope, request.required_scope); break;
    case "relations": {
      const items = rows(request.items);
      result = rows(request.gates).map(gate => items.map(item => todoGateRelation(gate, item)));
      break;
    }
    case "relation": result = todoGateRelation(requireJsonObject(request.gate, "gate"), requireJsonObject(request.item, "item")); break;
    case "scope_relation": result = decisionScopeRelation(requireJsonObject(request.gate, "gate"), requireJsonObject(request.item, "item")); break;
    case "exact_relation": result = exactTodoGateRelation(requireJsonObject(request.gate, "gate"), requireJsonObject(request.item, "item")); break;
    default: throw new TypeError("unsupported decision scope operation");
  }
  return {schema_version: "todo_decision_scope_result_v0", result};
}

File: loopx/control_plane/todos/decision_scope.py
"""Legacy input codec for the single typed decision-dependency rule owner."""
from __future__ import annotations

from typing import Any

from ..effect_runtime import effect_runtime_result
from .contract import (
    normalize_todo_blocks_agent,
    normalize_todo_claimed_by,
    normalize_todo_decision_scope,
    normalize_todo_decision_scope_outcomes,
    normalize_todo_global_gate,
    normalize_todo_id,
    normalize_todo_required_decision_scopes,
)
from .user_gate import is_user_gate_todo_item
from .projection import todo_projection_sort_key, todo_item_task_class, todo_item_has_removed_continuation_policy

TODO_GATE_BLOCKING_STATES = frozenset(
    {"gate_targets_todo", "gate_covers_action", "projection_repair_required"}
)
DECISION_SCOPE_CONSISTENCY_SCHEMA_VERSION = "required_decision_scope_consistency_v0"
STANDING_DECISION_AUTHORITY_SCHEMA_VERSION = "standing_decision_authority_v0"
DECISION_SCOPE_RELATION_SCHEMA_VERSION = "decision_scope_relation_v0"
TODO_GATE_RELATION_SCHEMA_VERSION = "todo_gate_relation_v0"
TYPED_DECISION_SCOPE_RESULT_SCHEMA_VERSION = "todo_decision_scope_result_v0"
_RELATION_SCHEMA_VERSIONS = frozenset(
    {DECISION_SCOPE_RELATION_SCHEMA_VERSION, TODO_GATE_RELATION_SCHEMA_VERSION}
)
_AGENT_SUMMARY_ITEM_KEYS = (
    "current_agent_claimed_open_items", "current_agent_claimed_advancement_items",
    "first_executable_items", "executable_backlog_items", "first_open_items", "backlog_items", "items",
)
_USER_SUMMARY_ITEM_KEYS = (
    "gate_open_items", "user_action_open_items", "first_open_items",
    "backlog_items", "items", "other_agent_scoped_items",
)


def _evaluate(operation: str, **facts: Any) -> object:
    result = effect_runtime_result("todo.decision_scope.evaluate", {
        "schema_version": "todo_decision_scope_request_v0", "operation": operation, **facts,
    })
    if (
        not isinstance(result, dict)
        or result.get("schema_version")
        != TYPED_DECISION_SCOPE_RESULT_SCHEMA_VERSION
        or "result" not in result
    ):
        raise TypeError("invalid typed decision scope projection")
    evaluated: object = result["result"]
    return evaluated


def _projection(
    operation: str,
    value: object,
    *,
    schema_versions: frozenset[str],
    nullable: bool,
) -> dict[str, Any] | None:
    if value is None and nullable:
        return None
    if not isinstance(value, dict) or any(
        not isinstance(key, str) for key in value
    ):
        raise TypeError(f"invalid typed decision scope {operation} projection")
    projection = {
        key: item for key, item in value.items() if isinstance(key, str)
    }
    if projection.get("schema_version") not in schema_versions:
        raise TypeError(f"invalid typed decision scope {operation} schema")
    return projection


def _required_projection(
    operation: str,
    value: object,
    *,
    schema_version: str,
) -> dict[str, Any]:
    projection = _projection(
        operation,
        value,
        schema_versions=frozenset({schema_version}),
        nullable=False,
    )
    if projection is None:  # Defensive narrowing; nullable=False rejects None.
        raise TypeError(f"invalid typed decision scope {operation} projection")
    return projection


def _optional_relation(operation: str, value: object) -> dict[str, Any] | None:
    return _projection(
        operation,
        value,
        schema_versions=_RELATION_SCHEMA_VERSIONS,
        nullable=True,
    )


def _relation_matrix(
    value: object,
    *,
    gate_count: int,
    item_count: int,
) -> list[list[dict[str, Any] | None]]:
    if not isinstance(value, list) or len(value) != gate_count:
        raise TypeError("invalid typed decision scope relations matrix")
    matrix: list[list[dict[str, Any] | None]] = []
    for row in value:
        if not isinstance(row, list) or len(row) != item_count:
            raise TypeError("invalid typed decision scope relations matrix")
        matrix.append([_optional_relation("relations", item) for item in row])
    return matrix


def _facts(item: dict[str, Any]) -> dict[str, Any]:
    return {
        "todo_id": normalize_todo_id(item.get("todo_id")),
        "status": str(item.get("status") or "open"), "done": item.get("done") is True,
        "claimed_by": normalize_todo_claimed_by(item.get("claimed_by")),
        "blocks_agent": normalize_todo_blocks_agent(item.get("blocks_agent")),
        "global_gate": bool(normalize_todo_global_gate(item.get("global_gate"))),
        "unblocks_todo_id": normalize_todo_id(item.get("unblocks_todo_id")),
        "decision_scope": normalize_todo_decision_scope(item.get("decision_scope")),
        "required_decision_scopes": normalize_todo_required_decision_scopes(item.get("required_decision_scopes")),
        "decision_scope_outcomes": normalize_todo_decision_scope_outcomes(item.get("decision_scope_outcomes")),
        # Preserve the existing legacy classification boundary, not new prose authority.
        "is_gate": is_user_gate_todo_item(item),
    }


def _source(summary: dict[str, Any] | None, source: list[dict[str, Any]] | None,
            keys: tuple[str, ...]) -> list[dict[str, Any]]:
    if source is not None:
        return [_facts(item) for item in source if isinstance(item, dict)]
    if not isinstance(summary, dict):
        return []
    result = []
    seen = set()
    for key in keys:
        values = (summary or {}).get(key)
        for item in values if isinstance(values, list) else []:
            if not isinstance(item, dict):
                continue
            if item.get("done") is True or str(item.get("status") or "open") not in {"open", "blocked"}:
                continue
            identity = (item.get("todo_id"), item.get("index"), item.get("text"))
            if identity not in seen:
                seen.add(identity)
                result.append(_facts(item))
    return result


def _authority(authority: dict[str, Any] | None) -> dict[str, Any] | None:
    if not isinstance(authority, dict) or not isinstance(authority.get("entries"), list):
        return None
    def decoded(entry: dict[str, Any]) -> dict[str, Any]:
        codecs = {"global_gate": normalize_todo_global_gate, "blocks_agent": normalize_todo_blocks_agent,
                  "claimed_by": normalize_todo_claimed_by, "decision_scope": normalize_todo_decision_scope}
        return {**entry, **{key: codec(entry[key]) for key, codec in codecs.items() if key in entry}}

    return {**authority, "entries": [decoded(entry) for entry in authority["entries"] if isinstance(entry, dict)],
            "conflicts": [decoded(entry) for entry in authority.get("conflicts", []) if isinstance(entry, dict)]
            if isinstance(authority.get("conflicts"), list) else []}


def standing_decision_authority_for_agent(authority: dict[str, Any] | None, *,
                                          agent_id: str | None) -> dict[str, Any] | None:
    normalized_authority = _authority(authority)
    if normalized_authority is None:
        return None
    return _projection(
        "standing",
        _evaluate(
            "standing",
            authority=normalized_authority,
            agent_id=normalize_todo_claimed_by(agent_id),
        ),
        schema_versions=frozenset({STANDING_DECISION_AUTHORITY_SCHEMA_VERSION}),
        nullable=True,
    )


def build_required_decision_scope_consistency(
    agent_todo_summary: dict[str, Any] | None, user_todo_summary: dict[str, Any] | None, *,
    agent_id: str | None, registered_agent_ids: list[str] | None = None,
    agent_source_items: list[dict[str, Any]] | None = None,
    user_source_items: list[dict[str, Any]] | None = None,
    standing_decision_authority: dict[str, Any] | None = None,
) -> dict[str, Any]:
    return _required_projection(
        "consistency",
        _evaluate(
            "consistency",
            agent_items=_source(
                agent_todo_summary,
                agent_source_items,
                _AGENT_SUMMARY_ITEM_KEYS,
            ),
            user_items=_source(
                user_todo_summary,
                user_source_items,
                _USER_SUMMARY_ITEM_KEYS,
            ),
            agent_id=normalize_todo_claimed_by(agent_id),
            registered_agents=sorted(
                {
                    value
                    for raw in registered_agent_ids or []
                    if (value := normalize_todo_claimed_by(raw))
                }
            ),
            standing_authority=_authority(standing_decision_authority),
        ),
        schema_version=DECISION_SCOPE_CONSISTENCY_SCHEMA_VERSION,
    )


def build_required_decision_scope_repair_hint(
    consistency: dict[str, Any],
) -> dict[str, Any] | None:
    if consistency.get("ok") is not False:
        return None
    raw_errors = consistency.get("errors")
    errors = raw_errors if isinstance(raw_errors, list) else []
    missing_gate_scope = any(
        isinstance(error, dict)
        and error.get("reason_code") == "multi_agent_user_gate_missing_scope"
        for error in errors
    )
    if missing_gate_scope:
        return {
            "source": "quota.should-run",
            "trigger": "user_gate_scope_projection_drift",
            "recommended_mode": "repair_user_gate_scope_projection",
            "effective_action": "todo_decision_scope_projection_repair",
            "blocked_action_scope": "todo_user_gate_scope_projection",
            "allowed": True,
            "notify": "DONT_NOTIFY",
            "reason": (
                "a multi-agent user_gate lacks blocks_agent or explicit "
                "global_gate=true, so its blocking authority is ambiguous"
            ),
            "repair_focus": (
                "set blocks_agent to one registered agent, set global_gate=true for "
                "intentional goal-wide authority, or downgrade the item to user_action"
            ),
            "spend_policy": (
                "spend once only after the gate-scope projection repair is validated "
                "and written back"
            ),
            "consistency": consistency,
        }
    result = {
        "source": "quota.should-run",
        "trigger": "required_decision_scope_projection_drift",
        "recommended_mode": "repair_required_decision_scope_projection",
        "effective_action": "todo_decision_scope_projection_repair",
        "blocked_action_scope": "todo_decision_scope_projection",
        "allowed": True,
        "notify": "DONT_NOTIFY",
        "reason": (
            "an agent todo requires a decision scope that does not resolve to a "
            "compatible open user_gate"
        ),
        "repair_focus": (
            "remove stale required_decision_scopes, create the explicit blocking "
            "user_gate, or correct its agent ownership; user_action remains non-blocking"
        ),
        "spend_policy": (
            "spend once only after the todo/gate projection repair is validated and written back"
        ),
        "consistency": consistency,
    }
    if any(isinstance(error, dict) and error.get("reason_code") == "standing_decision_order_unresolved"
           for error in errors):
        result.update(
            reason="contradictory standing decisions have no established chronological winner",
            repair_focus=(
                "reconcile decision chronology against explicit owner evidence; do not remove "
                "required_decision_scopes or invent approval merely to clear this conflict"
            ),
        )
    elif any(isinstance(error, dict) and error.get("reason_code") == "required_decision_scope_target_mismatch"
             for error in errors):
        result.update(
            reason="a scope-covering user gate explicitly targets a different Todo",
            repair_focus=(
                "reconcile the exact target and required scope against owner intent; do not "
                "remove the requirement, retarget the gate, or invent approval merely to clear the conflict"
            ),
        )
    return result


def decision_scope_covers(gate_scope: Any, required_scope: Any) -> bool:
    gate = normalize_todo_decision_scope(gate_scope)
    required = normalize_todo_decision_scope(required_scope)
    if not gate or not required:
        return False
    result = _evaluate("covers", gate_scope=gate, required_scope=required)
    if not isinstance(result, bool):
        raise TypeError("invalid typed decision scope covers projection")
    return result


def decision_scope_gate_relation(gate: dict[str, Any], agent_item: dict[str, Any]) -> dict[str, Any] | None:
    return _projection(
        "scope_relation",
        _evaluate("scope_relation", gate=_facts(gate), item=_facts(agent_item)),
        schema_versions=frozenset({DECISION_SCOPE_RELATION_SCHEMA_VERSION}),
        nullable=True,
    )


def exact_todo_gate_relation(gate: dict[str, Any], agent_item: dict[str, Any]) -> dict[str, Any] | None:
    return _projection(
        "exact_relation",
        _evaluate("exact_relation", gate=_facts(gate), item=_facts(agent_item)),
        schema_versions=frozenset({TODO_GATE_RELATION_SCHEMA_VERSION}),
        nullable=True,
    )


def todo_gate_relation(gate: dict[str, Any], agent_item: dict[str, Any]) -> dict[str, Any] | None:
    return _optional_relation(
        "relation",
        _evaluate("relation", gate=_facts(gate), item=_facts(agent_item)),
    )


def todo_gate_relations(gates: list[dict[str, Any]], items: list[dict[str, Any]]) -> list[list[dict[str, Any] | None]]:
    """Evaluate a consumer's candidate set in one RPC, retaining positional identity."""
    if not gates or not items:
        return [[] for _ in gates]
    return _relation_matrix(
        _evaluate(
            "relations",
            gates=[_facts(gate) for gate in gates],
            items=[_facts(item) for item in items],
        ),
        gate_count=len(gates),
        item_count=len(items),
    )


def todo_gate_relation_blocks_agent(relation: dict[str, Any] | None) -> bool:
    return bool(relation and relation.get("state") in TODO_GATE_BLOCKING_STATES)


def select_scoped_gate_fallback(gates: list[dict[str, Any]], items: list[dict[str, Any]], *,
                               agent_id: str | None, allow_unrelated_gate: bool,
                               monitor_debt_backoff_active: bool) -> dict[str, Any] | None:
    """Decode legacy facts; the typed owner returns positions, never display rows."""
    if not gates or not items:
        return None
    from .contract import normalize_todo_bound_agent, normalize_todo_excluded_agents

    def facts(item: dict[str, Any]) -> dict[str, Any]:
        priority, index = todo_projection_sort_key(item)
        return {**_facts(item), "action_kind": item.get("action_kind"),
                "archive_state": item.get("archive_state"), "resume_ready": item.get("resume_ready") is True,
                "bound_agent": normalize_todo_bound_agent(item.get("bound_agent")),
                "excluded_agents": normalize_todo_excluded_agents(item.get("excluded_agents")),
                "removed": todo_item_has_removed_continuation_policy(item),
                "task_class": todo_item_task_class(item), "priority_rank": priority, "persisted_index": index}

    return _projection(
        "fallback",
        _evaluate("fallback", gates=[facts(gate) for gate in gates], candidates=[facts(item) for item in items],
                  agent_id=normalize_todo_claimed_by(agent_id), allow_unrelated_gate=allow_unrelated_gate,
                  monitor_debt_backoff_active=monitor_debt_backoff_active),
        schema_versions=frozenset({"scoped_gate_fallback_selection_v0"}),
        nullable=True,
    )
